### ISSUE TITLE
LocalCollection.update doesn't remove dollar operators properly on upsert

### ISSUE THREAD

**Issue by jedwards1211** [meteor/meteor#5611]
Consider the following valid upsert that works in the Mongo shell:

```
meteor:PRIMARY> db.blah.update({name: {$exists: false}}, {$setOnInsert: {hello: 'world'}}, {upsert: true})
WriteResult({
    "nMatched" : 0,
    "nUpserted" : 1,
    "nModified" : 0,
    "_id" : ObjectId("563a9bba6f3a3efd390d7d20")
})
```

Mongo.Collection.upsert fails to do this correctly because it doesn't correctly remove the `$exists` operator:

```
> var blah = new Mongo.Collection('blah');
> blah.upsert({name: {$exists: false}}, {$setOnInsert: {hello: 'world'}});

MongoError: The dollar ($) prefixed field '$exists' in 'name.$exists' is not valid for storage.
    at Object.Future.wait (/home/andy/.meteor/packages/meteor-tool/.1.0.45.g1vjm8++os.linux.x86_64+web.browser+web.cordova/mt-os.linux.x86_64/dev_bundle/server-lib/node_modules/fibers/future.js:398:15)
    at [object Object].<anonymous> (/home/andy/jcore-portal/meteor/.meteor/local/build/programs/server/packages/meteor/helpers.js:119:1)
    at [object Object].MongoConnection.(anonymous function) [as update] (/home/andy/jcore-portal/meteor/.meteor/local/build/programs/server/packages/mongo/mongo_driver.js:678:1)
    at Object.fields (/home/andy/jcore-portal/meteor/.meteor/local/build/programs/server/packages/matb33:collection-hooks/update.js:75:1)
    at Object.collection.(anonymous function) [as update] (/home/andy/jcore-portal/meteor/.meteor/local/build/programs/server/packages/matb33:collection-hooks/collection-hooks.js:103:1)
    at [object Object].Mongo.Collection.(anonymous function) (/home/andy/jcore-portal/meteor/.meteor/local/build/programs/server/packages/mongo/collection.js:575:1)
    at [object Object].Mongo.Collection.(anonymous function) [as update] (/home/andy/jcore-portal/meteor/.meteor/local/build/programs/server/packages/aldeed:collection2/collection2.js:187:1)
    at [object Object].Mongo.Collection.upsert (/home/andy/jcore-portal/meteor/.meteor/local/build/programs/server/packages/mongo/collection.js:609:1)
    at repl:1:-56
    at /home/andy/jcore-portal/meteor/.meteor/local/build/programs/server/shell-server.js:244:23
    - - - - -
    at Object.toError (/home/andy/.meteor/packages/mongo/.1.1.0.zb7oxo++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/mongodb/utils.js:114:11)
    at /home/andy/.meteor/packages/mongo/.1.1.0.zb7oxo++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/mongodb/collection/core.js:577:27
    at /home/andy/.meteor/packages/mongo/.1.1.0.zb7oxo++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/mongodb/db.js:1195:7
    at /home/andy/.meteor/packages/mongo/.1.1.0.zb7oxo++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/mongodb/db.js:1903:9
    at Server.Base._callHandler (/home/andy/.meteor/packages/mongo/.1.1.0.zb7oxo++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/mongodb/connection/base.js:453:41)
    at /home/andy/.meteor/packages/mongo/.1.1.0.zb7oxo++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/mongodb/connection/server.js:487:18
    at [object Object].MongoReply.parseBody (/home/andy/.meteor/packages/mongo/.1.1.0.zb7oxo++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/mongodb/responses/mongo_reply.js:68:5)
    at [object Object].<anonymous> (/home/andy/.meteor/packages/mongo/.1.1.0.zb7oxo++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/mongodb/connection/server.js:445:20)
    at [object Object].emit (events.js:95:17)
    at [object Object].<anonymous> (/home/andy/.meteor/packages/mongo/.1.1.0.zb7oxo++os+web.browser+web.cordova/npm/node_modules/mongodb/lib/mongodb/connection/connection_pool.js:207:13)
```

The problem lies in [LocalCollection.update](https://github.com/meteor/meteor/blob/devel/packages/minimongo/minimongo.js#L735).  The following code for removing the dollar operators is incorrect because it only looks for dollar operators in the root of the selector:

``` js
var newDoc = LocalCollection._removeDollarOperators(selector);
```

Here is the failure to remove in Meteor shell:

``` js
> LocalCollection._removeDollarOperators({name: {$in: ['a', 'b']}})
{ name: { '$in': [ 'a', 'b' ] } }
```

And proof that it only takes effect on root keys:

``` js
> LocalCollection._removeDollarOperators({$in: ['a', 'b']})
{}
```

I'm not completely familiar with all possible query operators, but I'm assuming they can only occur as keys one level down, as in my example query.  In that case something like this should remove them properly (**EDIT** this is lodash `omit`, not underscore; I don't know if an equivalent exists in underscore):

``` js
var newDoc = _.omit(selector, function(valueSelector) { return isOperatorObject(valueSelector); });
```

**Comment by zol** [comment]
Thanks for reporting this @jedwards1211 . We'll take a closer look at this when we get a chance.

**Comment by sebakerckhof** [comment]
Okay, this should be fixed in next 1.6 build now that https://github.com/meteor/meteor/issues/8815 is merged.

**Comment by hwillson** [comment]
Awesome, thanks again for working on this @sebakerckhof.
